### ISSUE TITLE
Internal Server Error trying to get list of images

### ISSUE THREAD

**Issue by pavlovpetro** [brennerm/docker-registry-frontend#5]
Hello @brennerm 

I got the next issue when I was trying to get a list of images:

![error](https://user-images.githubusercontent.com/36917400/40712528-9ed7003a-6406-11e8-8097-378666864890.png)

Though I can see the repositories:

![name_of _repos](https://user-images.githubusercontent.com/36917400/40713754-62e636b0-6409-11e8-9d50-4eb9d11b6e6c.png)

"docker logs -f reg-ui" shows following:

```127.0.0.1 - - [30/May/2018 09:32:53] "GET /registry/registry-it/repo/devex36 HTTP/1.0" 500 -
[2018-05-30 09:37:17,166] ERROR in app: Exception on /registry/registry-it/repo/iisserver [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "frontend.py", line 137, in tag_overview
    repo=urldecode_filter(repo))
  File "/usr/local/lib/python3.6/site-packages/flask/templating.py", line 135, in render_template
    context, ctx.app)
  File "/usr/local/lib/python3.6/site-packages/flask/templating.py", line 117, in _render
    rv = template.render(context)
  File "/usr/local/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/root/templates/tag_overview.html", line 1, in top-level template code
    {% extends "layout.html" %}
  File "/root/templates/layout.html", line 21, in top-level template code
    <div id="content" class="container-fluid">{% block content %}{% endblock %}</div>
  File "/root/templates/tag_overview.html", line 21, in block "content"
    <td>{{ registry.get_size_of_layers(repo, tag) | to_mb }} MB</td>
  File "/root/docker_registry_frontend/registry.py", line 97, in get_size_of_layers
    result += self.get_size_of_layer(repo, image_id)
  File "/root/docker_registry_frontend/cache.py", line 21, in decorator
    result = f(*args, **kwargs)
  File "/root/docker_registry_frontend/registry.py", line 337, in get_size_of_layer
    method='HEAD'
  File "/root/docker_registry_frontend/registry.py", line 64, in request
    return urllib.request.urlopen(request, timeout=3)
  File "/usr/local/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/local/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/local/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/local/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
127.0.0.1 - - [30/May/2018 09:37:17] "GET /registry/registry-it/repo/iisserver HTTP/1.0" 500 - 
```

But "curl" returns the folowing:
`curl http://192.168.36.37:5000/v2/iisserver/tags/list`
{"name":"iisserver","tags":["base"]}

Docker version: 18.03.1-ce.

I was building an image from Dockerfile and running the prebuild image but I've got the same error in the both cases.
Could you help me? Thanks in advance.

**Comment by brennerm** [comment]
Hey @pavlovpetr,

I couldn't reproduce your problem on my side. It seems like the layer HTTP endpoint is not available or it doesn't support the HEAD method.

Which Docker registry version are you using?

**Comment by pavlovpetro** [comment]
Hi @brennerm,
Thanks for your quick answer!

I'm using version 2. 

Here is my config.yml(this registry has two roles - as a registry and as a proxy-cache.):
```
  1 version: 0.1
  2 log:
  3   accesslog:
  4     disable: true
  5   level: info
  6   formatter: text
  7   fields:
  8     service: registry
  9 storage:
 10   cache:
 11     blobdescriptor: inmemory
 12   filesystem:
 13     rootdirectory: /var/lib/registry
 14   maintenance:
 15      uploadpurging:
 16        enable: true
 17        age: 168h
 18        interval: 24h
 19        dryrun: true
 20 http:
 21   addr: :5000
 22   headers:
 23     X-Content-Type-Options: [nosniff]
 24 health:
 25   storagedriver:
 26     enabled: true
 27     interval: 5m
 28     threshold: 3
 29 proxy:
 30   remoteurl: http://192.168.32.67:5000
```

**Comment by brennerm** [comment]
Please adjust the following curl commands to your registry/repo/tag and execute them. My guess is that the second request will return a 404.

```
$ curl 172.17.0.2:5000/v2/ubuntu/manifests/16.04                                                                                                                                            
{                                                                                                                                                                                                                 
...                                                                                                                                                                         
   "fsLayers": [
      {
         "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
      },
      {
         "blobSum": "sha256:1a776d5f8f21442c36036f32af57c3bca06450718bc89623a42efd63442ab809"
      },
      {
         "blobSum": "sha256:94925a7a8f895b9bfae70aa0e767634353bb2514863af8c6d299749b70102f8d"
      },
      {
         "blobSum": "sha256:c7bccbb1d18371fe9f6514c1d93f9c8007b4653a0d22aadbc3ed56c8faf840a6"
      },
      {
         "blobSum": "sha256:4791a9f808607abc64b0398fb4a88b7e79d44a770ca0c7206cd8b0d12e307e02"
      },                                                                                                                                                                                                          
      {                                                                                                                                                                                                           
         "blobSum": "sha256:4f1bb8b6572003ca3526ab23c80c385c5eadd1748e3996e444b66d47ce0ee7af"
      }
   ],              
...
}

# replace hash with one from the previous result
$ curl --head "172.17.0.2:5000/v2/ubuntu/blobs/sha256:1a776d5f8f21442c36036f32af57c3bca06450718bc89623a42efd63442ab809"

HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: max-age=31536000
Content-Length: 169
Content-Type: application/octet-stream
Docker-Content-Digest: sha256:1a776d5f8f21442c36036f32af57c3bca06450718bc89623a42efd63442ab809
Docker-Distribution-Api-Version: registry/2.0
Etag: "sha256:1a776d5f8f21442c36036f32af57c3bca06450718bc89623a42efd63442ab809"
Date: Wed, 30 May 2018 12:17:06 GMT
```

**Comment by pavlovpetro** [comment]
No, it works

```
curl 192.168.36.37:5000/v2/mssql2017/manifests/base
{
   "schemaVersion": 1,
   "name": "mssql2017",
   "tag": "base",
   "architecture": "amd64",
   "fsLayers": [
      {
         "blobSum": "sha256:b3e281759826440e1f0a0be19cfafd42313c5f4c8d64d13d06a15b61093a88b2"
      },
      {
         "blobSum": "sha256:ae251644e9754318b9ae375fcda97b4ea17f807e2a9bf94d27fc7a6027439e6f"
      },
      {
         "blobSum": "sha256:69bb00a82fa22d8ee1f3be7666b373e7749f697a8642288734623e244927d436"
      },
      {
         "blobSum": "sha256:42c0168ffd332ffe151815269f87a28c83ca741b41e301c594a7edb3ab9f9b74"
      },
      {
         "blobSum": "sha256:1281c54801fa02a184d6d3594619b9face960fd070a861e979fa42f1fb5c2c81"
      },
      {
         "blobSum": "sha256:b04d8f79c3b811dfde568a3f9b3921eaa3838f3876e9cf04687701065a5b3220"
      },
      {
         "blobSum": "sha256:24e28090e8e9cf978d4a315a6b2095d8c8614d3d7da2314b5ab4ad654e2e8bfe"
      },
      {
         "blobSum": "sha256:6233756d5d259762493102827a1463ae8968f440123503db7c4e4319952701cf"
      },
      {
         "blobSum": "sha256:9d3e236e32bbb3aa8bfcbab22fe6a031d37d5792581756eb03a9b6b29a0dd46c"
      },
      {
         "blobSum": "sha256:cfb27c9ba25f60372361ea8779c927f066c385b6339e29fda5c739feb3163686"
      },
      {
         "blobSum": "sha256:3889bb8d808bbae6fa5a33e07093e65c31371bcf9e4c38c21be6b9af52ad1548"
      }
   ],

curl --head "192.168.36.37:5000/v2/mssql2017/blobs/sha256:ae251644e9754318b9ae375fcda97b4ea17f807e2a9bf94d27fc7a6027439e6f"


HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: max-age=31536000
Content-Length: 8600320
Content-Type: application/octet-stream
Docker-Content-Digest: sha256:ae251644e9754318b9ae375fcda97b4ea17f807e2a9bf94d27fc7a6027439e6f
Docker-Distribution-Api-Version: registry/2.0
Etag: "sha256:ae251644e9754318b9ae375fcda97b4ea17f807e2a9bf94d27fc7a6027439e6f"
X-Content-Type-Options: nosniff
Date: Wed, 30 May 2018 12:32:54 GMT
```

**Comment by brennerm** [comment]
Does the error occur for every repo or just for the iisserver?

**Comment by pavlovpetro** [comment]
Unfortunately for all repos in all my registries. I have two.
It occurs when I try to get list of images only.
Lists of registries and repos returns without any issues.

**Comment by brennerm** [comment]
Even using your registry configuration I can't reproduce the problem. 

Which registry version are you using precisely? 2.5, 2.6?

**Comment by pavlovpetro** [comment]
I'm using v2.6.2. I cannot understand why it doesn't work... 
I've turned on debugger but it showed  nothing. All variables get correct data(hash sums of repos, digest name of tags etc.)... Network settings are correct. Why it returns 404, what it cannot find?

**Comment by pavlovpetro** [comment]
Oh, I figured out, unfortunately 'docker-registry-frontend' doesn't work/support the images which are built on windows core... Could you please fix it?

**Comment by brennerm** [comment]
Fixing this is gonna be tough as I don't have any experience/access with/to Docker running on Windows.

Having a look at your comment I don't see any difference in the HTTP response you get from your registry.

**Comment by pavlovpetro** [comment]
I've put an ubuntu image to the registry which is running in Docker on Windows and the 'docker-registry-frontend' works well with this image. But with the images which contain Windows Core doesn't work on the same registry. The similar situation with the registry that is running on Linux.  I think the solution is somewhere in the Flask but not sure.(probably,need enable encode/decode for Windows)

**Comment by pavlovpetro** [comment]
It seems I understood why it doesn't work with Windows Images.
Using debugger I noticed all variables contain the latest value "blobsum" of the image.

![blobsum](https://user-images.githubusercontent.com/36917400/40780205-3d116eaa-64e0-11e8-8845-f11011294f3d.png)

![blobsum2](https://user-images.githubusercontent.com/36917400/40780236-554ec4e0-64e0-11e8-974c-e571a99e8a8b.png)

And when I tried to get a Header I got 404(as you supposed yesterday):

![blobsum3](https://user-images.githubusercontent.com/36917400/40780304-94246292-64e0-11e8-9984-e16240d4377f.png)

All images which are built on Windows contains these two latest blobsums. They are the base layers of all Windows images.
But the layers themselves are missed because they aren't kept on my local registry! They are kept on Docker Hub. And it is the cause I get 404.

I think if the variable "layer_id" get the first blobsum, it would solve the issue.  What do you think?

**Comment by brennerm** [comment]
Great investigation work, thank you. 

My proposal would be to just ignore 404's for this kind of requests. Just taking the first one wouldn't work as we need to iterate through all layers to determine the overall size of the image.

I'll have a look into this over the weekend.

**Comment by pavlovpetro** [comment]
I found a temporary solve for me - it is discarding from the "set" layer_ids not working blobsums.
file "manifest.py":

```
def get_layer_ids(self):
        layer_ids = []

        for layer in self._content['fsLayers']:
            layer_ids.append(layer['blobSum'])

        s=set(layer_ids)
        s.discard("sha256:3889bb8d808bbae6fa5a33e07093e65c31371bcf9e4c38c21be6b9af52ad1548")
        s.discard("sha256:449343c9d7e2919413898dc8a7e8780ef164b76a3b9dd19de104706edf05113a")
        s.discard("sha256:ead9f4ead3c5a712cb213a5318f4a8bf3604bc16e16ce4f7cf0b66f7d2073282")
        s.discard("sha256:cfb27c9ba25f60372361ea8779c927f066c385b6339e29fda5c739feb3163686")
        layer_ids = s
        return set(layer_ids)
```

**Comment by brennerm** [comment]
Please give it another try with the new changes in the support-windows-images branch.

**Comment by brennerm** [comment]
Solved with #6.
